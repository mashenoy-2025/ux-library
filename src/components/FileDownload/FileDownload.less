:root {
  /* re-use from design system or override if necessary */
  --table-background-color: var(--primary-background-color);
  --table-foreground-color: var(--primary-foreground-color);
  --table-border-color: #ccc;
  --table-header-font-weight: var(--font-weight-semibold);
  --table-cell-padding: var(--spacing-medium);
  --table-selected-row-background-color: #eee;
}

.FileDownload {
  width: 100%;
  overflow-x: auto;

  .FileDownload__toolbar {
    display: flex;
    align-items: center;
    padding: 0 var(--spacing-small);
    gap: 20px;
    margin: var(--spacing-small);
  }

  .FileDownload__table {
    width: 100%;
    border-collapse: collapse;
    background-color: var(--table-background-color);
    border: 1px solid var(--table-border-color);
    min-width: var(--mobile-breakpoint);

    th {
      background-color: var(--table-background-color);
      color: var(--table-foreground-color);
      font-weight: var(--table-header-font-weight);
      padding: var(--table-cell-padding);
      text-align: left;
      border-bottom: 1px solid var(--table-border-color);
    }

    td {
      padding: var(--table-cell-padding);
      color: var(--table-foreground-color);
      border-bottom: 1px solid var(--table-border-color);
    }

    tbody {
      tr {
        &.selected,
        &:hover {
          background-color: var(--table-selected-row-background-color);
          cursor: pointer;
        }

        &.disabled {
          cursor: not-allowed;
        }
      }
    }

    input[type="checkbox"] {
      padding: var(--spacing-medium);
    }
  }
}
