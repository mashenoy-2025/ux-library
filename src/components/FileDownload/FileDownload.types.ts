export interface IFileDownloadItem {
  id: string;
  name: string;
}

export interface IFileDownloadColumnType<T> {
  key: string;
  header: string;
  customRenderFunction?: (item: T) => React.ReactNode;
}

export type IFileDownloadSelectionMode = "single" | "multiple";

export interface IAction<T> {
  buttonText: string;
  onClick: (items: T[]) => void;
}

export interface IFileDownloadProps<T extends IFileDownloadItem> {
  items: T[];
  columns: IFileDownloadColumnType<T>[];
  // If we want to support multiple actions, switch to IAction<T>[] array.
  downloadActionProps: IAction<T>; //
  caption?: string;
  className?: string;

  // todo: Support selectionMode to allow single or multiple file downloads
  selectionMode: IFileDownloadSelectionMode;
  canDownloadFile: (item: T) => boolean;
}
