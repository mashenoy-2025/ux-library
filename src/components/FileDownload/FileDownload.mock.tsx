import * as React from "react";

import { IFileDownloadItem, IFileDownloadProps } from "./FileDownload.types";

export interface FraudInvestigationFile {
  id: string;
  name: string;
  device: string;
  path: string;
  status: "Scheduled" | "Available";
}

const files: FraudInvestigationFile[] = [
  {
    id: "3564d56c-dee1-47e3-a1d3-65252096c1bf",
    name: "smss.exe",
    device: "Mario",
    path: "\\Device\\HarddiskVolume2\\Windows\\System32\\smss.exe",
    status: "Scheduled",
  },

  {
    id: "bd817d8e-f1ce-4ccc-ba5a-cef2774e1179",
    name: "netsh.exe",
    device: "Luigi",
    path: "\\Device\\HarddiskVolume2\\Windows\\System32\\netsh.exe",
    status: "Available",
  },

  {
    id: "47e139d8-48bb-4461-8fec-c8a7886cf386",
    name: "uxtheme.dll",
    device: "Peach",
    path: "\\Device\\HarddiskVolume1\\Windows\\System32\\uxtheme.dll",
    status: "Available",
  },

  {
    id: "829d8a70-6df8-4419-a4e9-ccc02b1a2fde",
    name: "aries.sys",
    device: "Daisy",
    path: "\\Device\\HarddiskVolume1\\Windows\\System32\\aries.sys",
    status: "Scheduled",
  },

  {
    id: "969f5965-3118-43c0-9d6a-d450e582cf82",
    name: "cryptbase.dll",
    device: "Yoshi",
    path: "\\Device\\HarddiskVolume1\\Windows\\System32\\cryptbase.dll",
    status: "Scheduled",
  },

  {
    id: "5",
    name: "7za.exe",
    device: "Toad",
    path: "\\Device\\HarddiskVolume1\\temp\\7za.exe",
    status: "Scheduled",
  },
];

export type FileWithKey = FraudInvestigationFile & IFileDownloadItem;
const mockItems: FileWithKey[] = files.map((file) => ({
  ...file,
  key: file.id,
}));

export const mockFileDownloadProps: IFileDownloadProps<FileWithKey> = {
  items: mockItems,
  caption: "Download files for fraud analysis",
  downloadActionProps: {
    buttonText: "Download selected",

    // Parent/Consumer component handles the responsibility of how to download files.
    onClick: (items: FileWithKey[]) =>
      alert(items.map((i) => `${i.device} => ${i.path} \n`).join("\n")),
  },
  columns: [
    { header: "Name", key: "name" },
    { header: "Device", key: "device" },
    { header: "Path", key: "path" },
    {
      header: "Status",
      key: "status",
      customRenderFunction: ({ status }: FileWithKey) => (
        <div className="flex-row-centered">
          {status === "Available" && (
            <span role="presentation" className="status-green" />
          )}

          <span>{status}</span>
        </div>
      ),
    },
  ],
  canDownloadFile: ({ status }: FileWithKey) => status === "Available",
  selectionMode: "multiple",
};
