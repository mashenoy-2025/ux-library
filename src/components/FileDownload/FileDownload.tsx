import "./FileDownload.less";

import * as React from "react";

import {
  IAction,
  IFileDownloadItem,
  IFileDownloadProps,
} from "./FileDownload.types";
import { useCallback, useEffect, useMemo, useRef, useState } from "react";

import { classNames } from "../shared/utils";

export const FileDownload = <T extends IFileDownloadItem>({
  columns,
  items,
  downloadActionProps,
  caption,
  canDownloadFile,
}: IFileDownloadProps<T>) => {
  const [selectedItems, setSelectedItems] = useState<T[]>([]);
  const selectAllCheckboxRef = useRef<HTMLInputElement>(null);
  const { buttonText, onClick } = downloadActionProps;

  const isItemSelected = useCallback(
    (item: T) => selectedItems.some((s) => s.id === item.id),
    [selectedItems]
  );

  const filesAvailableForDownload = useMemo(
    () => items.filter(canDownloadFile),
    [canDownloadFile, items]
  );

  const isAllItemsSelected =
    selectedItems.length === filesAvailableForDownload.length;

  useEffect(() => {
    if (!selectAllCheckboxRef.current) return;

    selectAllCheckboxRef.current.indeterminate =
      selectedItems.length > 0 &&
      selectedItems.length < filesAvailableForDownload.length;
  }, [selectedItems]);

  const onItemSelectionToggle = useCallback(
    (item: T) => () => {
      if (!canDownloadFile(item)) {
        return;
      }

      const isItemAlreadySelected = selectedItems.some(
        ({ id }) => id === item.id
      );

      if (isItemAlreadySelected) {
        setSelectedItems([...selectedItems.filter(({ id }) => id !== item.id)]);
      } else {
        setSelectedItems([...selectedItems, item]);
      }
    },
    [selectedItems, setSelectedItems]
  );

  const onSelectAllToggle = useCallback(() => {
    if (isAllItemsSelected) {
      setSelectedItems([]);
    } else {
      setSelectedItems([...filesAvailableForDownload]);
    }
  }, [setSelectedItems, filesAvailableForDownload, isAllItemsSelected]);

  const onDownloadClick = useCallback(
    (onClick: IAction<T>["onClick"]) => () => {
      onClick(selectedItems);
    },
    [selectedItems]
  );

  const toolbarJsx = (
    <div className="FileDownload__toolbar">
      <span className="selection_information" aria-live="polite">
        {selectedItems.length
          ? `Selected: ${selectedItems.length}`
          : "None selected"}
      </span>

      <button
        type="button"
        onClick={onDownloadClick(onClick)}
        disabled={!selectedItems.length}
      >
        {buttonText}
      </button>
    </div>
  );

  return (
    <div className="FileDownload">
      {toolbarJsx}

      <table className="FileDownload__table">
        {caption && <caption className="sr-only">{caption}</caption>}
        <thead>
          <tr>
            <th key="header_selectAll">
              <input
                id="checkbox_selectAll"
                type="checkbox"
                onChange={onSelectAllToggle}
                checked={isAllItemsSelected}
                ref={selectAllCheckboxRef}
                aria-label="Select all files for download"
              />
            </th>

            {columns.map(({ key, header }) => (
              <th key={`header_${key}`} scope="col">
                {header}
              </th>
            ))}
          </tr>
        </thead>

        <tbody>
          {items.map((item) => (
            <tr
              key={item.id}
              className={classNames(
                isItemSelected(item) && "selected",
                !canDownloadFile(item) && "disabled"
              )}
              onClick={onItemSelectionToggle(item)}
            >
              <td key="selection">
                <input
                  id={`checkbox_select_${item.id}`}
                  type="checkbox"
                  checked={isItemSelected(item)}
                  disabled={!canDownloadFile(item)}
                  aria-label={`Select file ${item.name}`}
                  onChange={onItemSelectionToggle(item)}
                />
              </td>

              {columns.map(({ key, customRenderFunction }) => (
                <td key={key}>
                  {customRenderFunction
                    ? customRenderFunction(item)
                    : item[key]}
                </td>
              ))}
            </tr>
          ))}

          {/* todo: Handle no items scenario */}
        </tbody>
      </table>
    </div>
  );
};
