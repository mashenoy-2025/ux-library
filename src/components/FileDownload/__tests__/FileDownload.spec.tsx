import {} from "@testing-library/user-event";

import * as React from "react";

import { FileWithKey, mockFileDownloadProps } from "../FileDownload.mock";
import { render, screen } from "@testing-library/react";

import { FileDownload } from "../FileDownload";
import { IFileDownloadProps } from "../FileDownload.types";
import userEvent from "@testing-library/user-event";

const scheduledFile = mockFileDownloadProps.items[0];
const availableFile = mockFileDownloadProps.items[1];
const user = userEvent.setup();

describe("FileDownload", () => {
  let defaultProps: IFileDownloadProps<FileWithKey>;

  beforeEach(() => {
    defaultProps = { ...mockFileDownloadProps };
  });

  it("should match default snapshot", () => {
    expect(
      render(<FileDownload {...defaultProps} />).container
    ).toMatchSnapshot();
  });

  it(`should not allow download for scheduled state`, () => {
    render(<FileDownload {...defaultProps} />);

    expect(
      screen.getByLabelText<HTMLInputElement>(
        `Select file ${scheduledFile.key}`
      )
    ).toBeDisabled();
  });

  it("should download selected files on download button click", async () => {
    const downloadClickFn = vi.fn();
    const props = {
      ...defaultProps,
      downloadActionProps: {
        ...defaultProps.downloadActionProps,
        onClick: downloadClickFn,
      },
    };

    render(<FileDownload {...props} />);

    // Download available file
    await user.click(screen.getByLabelText(`Select file ${availableFile.key}`));
    await user.click(screen.getByText("Download selected"));
    expect(downloadClickFn).toHaveBeenCalledWith([availableFile]);
  });

  describe("validate select all checkbox states", () => {
    it("select all", async () => {
      render(<FileDownload {...defaultProps} />);

      const selectAllCheckbox = screen.getByLabelText(
        "Select all files for download"
      );

      expect(screen.getByText("None selected")).toBeInTheDocument();
      await user.click(selectAllCheckbox);
      expect(screen.getByText("Selected: 2")).toBeInTheDocument();
    });

    it("select all - indeterminate state", async () => {
      render(<FileDownload {...defaultProps} />);

      const selectAllCheckbox = screen.getByLabelText(
        "Select all files for download"
      );

      await user.click(selectAllCheckbox);
      expect(screen.getByText("Selected: 2")).toBeInTheDocument();
    });

    it("select all - indeterminate", async () => {
      render(<FileDownload {...defaultProps} />);

      await user.click(
        screen.getByLabelText(`Select file ${availableFile.key}`)
      );

      const selectAllCheckbox = screen.getByLabelText<HTMLInputElement>(
        "Select all files for download"
      );

      expect(screen.getByText("Selected: 1")).toBeInTheDocument();
      expect(selectAllCheckbox.indeterminate).toBe(true);
    });
  });

  it.todo("should validate a11y fast pass checks", () => {
    // Test for accessibility errors using axe core library
  });
});
