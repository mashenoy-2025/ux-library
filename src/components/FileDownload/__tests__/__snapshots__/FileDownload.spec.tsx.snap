// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileDownload > should match default snapshot 1`] = `
<div>
  <div
    class="FileDownload"
  >
    <div
      class="FileDownload__toolbar"
    >
      <span
        class="selection_information"
      >
        None selected
      </span>
      <button
        disabled=""
        type="button"
      >
        Download selected
      </button>
    </div>
    <table
      class="FileDownload__table"
    >
      <caption
        class="sr-only"
      >
        Download files for fraud analysis
      </caption>
      <thead>
        <tr>
          <th
            scope="col"
          >
            <input
              aria-label="Select all files for download"
              id="checkbox_selectAll"
              type="checkbox"
            />
          </th>
          <th
            scope="col"
          >
            Name
          </th>
          <th
            scope="col"
          >
            Device
          </th>
          <th
            scope="col"
          >
            Path
          </th>
          <th
            scope="col"
          >
            Status
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="disabled"
        >
          <td
            scope="col"
          >
            <input
              aria-label="Select file 3564d56c-dee1-47e3-a1d3-65252096c1bf"
              disabled=""
              id="checkbox_select_3564d56c-dee1-47e3-a1d3-65252096c1bf"
              type="checkbox"
            />
          </td>
          <td>
            smss.exe
          </td>
          <td>
            Mario
          </td>
          <td>
            \\Device\\HarddiskVolume2\\Windows\\System32\\smss.exe
          </td>
          <td>
            <div
              class="flex-row-centered"
            >
              <span>
                Scheduled
              </span>
            </div>
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            scope="col"
          >
            <input
              aria-label="Select file bd817d8e-f1ce-4ccc-ba5a-cef2774e1179"
              id="checkbox_select_bd817d8e-f1ce-4ccc-ba5a-cef2774e1179"
              type="checkbox"
            />
          </td>
          <td>
            netsh.exe
          </td>
          <td>
            Luigi
          </td>
          <td>
            \\Device\\HarddiskVolume2\\Windows\\System32\\netsh.exe
          </td>
          <td>
            <div
              class="flex-row-centered"
            >
              <span
                class="status-green"
                role="presentation"
              />
              <span>
                Available
              </span>
            </div>
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            scope="col"
          >
            <input
              aria-label="Select file 47e139d8-48bb-4461-8fec-c8a7886cf386"
              id="checkbox_select_47e139d8-48bb-4461-8fec-c8a7886cf386"
              type="checkbox"
            />
          </td>
          <td>
            uxtheme.dll
          </td>
          <td>
            Peach
          </td>
          <td>
            \\Device\\HarddiskVolume1\\Windows\\System32\\uxtheme.dll
          </td>
          <td>
            <div
              class="flex-row-centered"
            >
              <span
                class="status-green"
                role="presentation"
              />
              <span>
                Available
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="disabled"
        >
          <td
            scope="col"
          >
            <input
              aria-label="Select file 829d8a70-6df8-4419-a4e9-ccc02b1a2fde"
              disabled=""
              id="checkbox_select_829d8a70-6df8-4419-a4e9-ccc02b1a2fde"
              type="checkbox"
            />
          </td>
          <td>
            aries.sys
          </td>
          <td>
            Daisy
          </td>
          <td>
            \\Device\\HarddiskVolume1\\Windows\\System32\\aries.sys
          </td>
          <td>
            <div
              class="flex-row-centered"
            >
              <span>
                Scheduled
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="disabled"
        >
          <td
            scope="col"
          >
            <input
              aria-label="Select file 969f5965-3118-43c0-9d6a-d450e582cf82"
              disabled=""
              id="checkbox_select_969f5965-3118-43c0-9d6a-d450e582cf82"
              type="checkbox"
            />
          </td>
          <td>
            cryptbase.dll
          </td>
          <td>
            Yoshi
          </td>
          <td>
            \\Device\\HarddiskVolume1\\Windows\\System32\\cryptbase.dll
          </td>
          <td>
            <div
              class="flex-row-centered"
            >
              <span>
                Scheduled
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="disabled"
        >
          <td
            scope="col"
          >
            <input
              aria-label="Select file 5"
              disabled=""
              id="checkbox_select_5"
              type="checkbox"
            />
          </td>
          <td>
            7za.exe
          </td>
          <td>
            Toad
          </td>
          <td>
            \\Device\\HarddiskVolume1\\temp\\7za.exe
          </td>
          <td>
            <div
              class="flex-row-centered"
            >
              <span>
                Scheduled
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
